src/bbs2gh/Services/BbsSmbArchiveDownloader.cs

```diff
@@ -0,0 +1,227 @@
+            }
+
+            return targetExportArchiveFullPath;
+        }
+        finally
+        {
+            if (sourceExportArchiveHandle != null)
+            {
+                fileStore?.CloseFile(sourceExportArchiveHandle);
+            }
+
+            fileStore?.Disconnect();
+            _smbClient.Logoff();
+            _smbClient.Disconnect();
+        }
+    }
+
+    private void LogProgress(long downloadedBytes, long? totalBytes)
+    {
+        if (DateTime.Now < _nextProgressReport)
+        {
+            return;
+        }
+
+        var totalProgressMessage = totalBytes.HasValue
+            ? $" out of {GetLogFriendlySize(totalBytes.Value)} ({GetPercentage(downloadedBytes, totalBytes.Value)})"
+            : "";
+        _log.LogInformation($"Download archive in progress, {GetLogFriendlySize(downloadedBytes)}{totalProgressMessage} completed...");
+
+        _nextProgressReport = _nextProgressReport.AddSeconds(DOWNLOAD_PROGRESS_REPORT_INTERVAL_IN_SECONDS);
+    }
+
+    private string GetPercentage(long downloadedBytes, long totalBytes)
+    {
+        if (totalBytes is 0L)
+        {
+            return "unknown%";
+        }
+
+        var percentage = (int)(downloadedBytes * 100D / totalBytes);
+        return $"{percentage}%";
+    }
+
+    private string GetLogFriendlySize(long size)
+    {
+        const int kilobyte = 1024;
+        const int megabyte = 1024 * kilobyte;
+        const int gigabyte = 1024 * megabyte;
+
+        return size switch
+        {
+            < kilobyte => $"{size:n0} bytes",
+            < megabyte => $"{size / (double)kilobyte:n0} KB",
+            < gigabyte => $"{size / (double)megabyte:n0} MB",
+            _ => $"{size / (double)gigabyte:n2} GB"
+        };
+    }
+
+    private FileStream OpenWriteTargetExportArchive(string targetExportArchiveFullPath)
+    {
+        if (_fileSystemProvider.FileExists(targetExportArchiveFullPath))
+        {
+            throw new OctoshiftCliException($"Target export archive ({targetExportArchiveFullPath}) already exists.");
+        }
+
+        _fileSystemProvider.CreateDirectory(Path.GetDirectoryName(targetExportArchiveFullPath));
+        return _fileSystemProvider.Open(targetExportArchiveFullPath, FileMode.CreateNew);
+    }
+
+    private void ConnectToHost()
+    {
+        var isConnected = IPAddress.TryParse(_host, out var ipAddress)
+            ? _smbClient.Connect(ipAddress, SMBTransportType.DirectTCPTransport)
+            : _smbClient.Connect(_host, SMBTransportType.DirectTCPTransport);
+
+        if (!isConnected)
+        {
+            throw new OctoshiftCliException($"Unable to connect to host \"{_host}\".");
+        }
+    }
+
+    private void Login()
+    {
+        var status = _smbClient.Login(_domainName ?? "", _smbUser, _smbPassword);
+
+        if (!IsSuccessStatus(status))
+        {
+            throw new OctoshiftCliException($"Unable to login with provided credentials (Status Code: {status}).");
```

Comment:
Same question here - should we make it clear that it's the SMB login that has failed rather than, for example, the login to the BBS API?

Reply:
Same thing here too, it would be perfectly clear that this happened during the SMB download.